### ISSUE TITLE
Non-null assertion operator non working on properties

### ISSUE THREAD

**Issue by HolgerJeromin** [Microsoft/TypeScript#15796]
**TypeScript Version:**  2.3

**Code**

```ts
export interface Dictionary<T> {
    [index: string]: T | undefined;
}
const name = 'bar';
let foo: Dictionary<number> = {};
if (foo[name] === undefined) {
    foo[name] = 1;
} else {
    let numberOrUndef = foo[name];
    let numb = foo[name]!;
    foo[name]++;
}
```

last code line `foo[name]++;` is probably undefined. Ok, it is not, but no problem.
But i cannot override this with the Non-null assertion operator !

`foo[name]!++;` gives "The operand of an increment or decrement operator must be a variable or a property access."

**Comment by mhegazy** [comment]
`checkReferenceExpression` should skip the `NonNullExpression` just like it does with parenthesized expressions.

**Comment by HolgerJeromin** [comment]
Ah, thanks for the hint with parenthesis as a workaround.

**Comment by mhegazy** [comment]
Parenthesis will not work. We already skip these correctelly and will land on the null assertion again and flag it as an error.

**Comment by sandersn** [comment]
Fix is up at #16465
